bpf/lib/nat.h

```diff
@@ -0,0 +1,872 @@
+
+	ostate = snat_v4_lookup(otuple);
+	if (ostate) {
+		snat_v4_swap_tuple(otuple, rtuple);
+		rtuple->daddr = ostate->to_saddr;
+		rtuple->dport = ostate->to_sport;
+	}
+
+	return ostate ? 0 : -1;
+}
+
+static __always_inline void snat_v4_ct_canonicalize(struct ipv4_ct_tuple *otuple)
+{
+	__be32 addr = otuple->saddr;
+
+	otuple->flags = NAT_DIR_EGRESS;
+	/* Workaround #5848. */
+	otuple->saddr = otuple->daddr;
+	otuple->daddr = addr;
+}
+
+static __always_inline void snat_v4_delete_tuples(struct ipv4_ct_tuple *otuple)
+{
+	struct ipv4_ct_tuple rtuple;
+
+	if (otuple->flags & TUPLE_F_IN)
+		return;
+	snat_v4_ct_canonicalize(otuple);
+	if (!snat_v4_reverse_tuple(otuple, &rtuple))
+		snat_v4_delete(otuple, &rtuple);
+}
+
+static __always_inline int snat_v4_new_mapping(struct __sk_buff *skb,
+					       struct ipv4_ct_tuple *otuple,
+					       struct ipv4_nat_entry *ostate)
+{
+	struct ipv4_nat_entry rstate;
+	struct ipv4_ct_tuple rtuple;
+	int ret, retries;
+
+	__builtin_memset(&rstate, 0, sizeof(rstate));
+	__builtin_memset(ostate, 0, sizeof(*ostate));
+
+	rstate.to_daddr = otuple->saddr;
+	rstate.to_dport = otuple->sport;
+
+	ostate->to_saddr = SNAT_IPV4_EXTERNAL;
+	ostate->to_sport = otuple->sport;
+
+	snat_v4_swap_tuple(otuple, &rtuple);
+	rtuple.daddr = SNAT_IPV4_EXTERNAL;
+
+	if (otuple->saddr == SNAT_IPV4_EXTERNAL) {
+		ostate->common.host_local = 1;
+		rstate.common.host_local = ostate->common.host_local;
+	}
+
+#pragma unroll
+	for (retries = 0; retries < SNAT_COLLISION_RETRIES; retries++) {
+		if (!snat_v4_lookup(&rtuple)) {
+			ostate->common.created = bpf_ktime_get_nsec();
+			rstate.common.created = ostate->common.created;
+
+			ret = snat_v4_update(otuple, ostate, &rtuple, &rstate);
+			if (!ret)
+				return 0;
+		}
+
+		rtuple.dport = ostate->to_sport = bpf_htons(
+			__snat_get_port_range(SNAT_MAPPING_MIN_PORT,
+					      SNAT_MAPPING_MAX_PORT));
+	}
+
+	return DROP_NAT_NO_MAPPING;
+}
+
+static __always_inline int snat_v4_track_local(struct __sk_buff *skb,
+					       struct ipv4_ct_tuple *tuple,
+					       struct ipv4_nat_entry *state,
+					       int dir, __u32 off)
+{
+	struct ct_state ct_state;
+	struct ipv4_ct_tuple tmp;
+	bool needs_ct = false;
+	__u32 monitor = 0;
+	int ret, where;
+
+	if (state && state->common.host_local) {
+		needs_ct = true;
+	} else if (!state && dir == NAT_DIR_EGRESS) {
+		if (tuple->saddr == SNAT_IPV4_EXTERNAL)
+			needs_ct = true;
+	}
+	if (!needs_ct)
+		return 0;
+
+	__builtin_memset(&ct_state, 0, sizeof(ct_state));
+	__builtin_memcpy(&tmp, tuple, sizeof(tmp));
+
+	where = dir == NAT_DIR_INGRESS ? CT_INGRESS : CT_EGRESS;
+
+	ret = ct_lookup4(get_ct_map4(&tmp), &tmp, skb, off, where,
+			 &ct_state, &monitor);
+	if (ret < 0) {
+		return ret;
+	} else if (ret == CT_NEW) {
+		ret = ct_create4(get_ct_map4(&tmp), &tmp, skb, where,
+				 &ct_state);
+		if (IS_ERR(ret))
+			return ret;
+	}
+
+	return 0;
+}
+
+static __always_inline int snat_v4_handle_mapping(struct __sk_buff *skb,
```

Comment:
This function would really benefit from having a brief comment above describing the inputs and expected output possibilities.

In particular, the case where there's no corresponding NAT state and it's ingress where we just pass the packet (presumably so that the stack can just attempt to receive it directly) is a bit difficult to decipher.

Reply:
Yeah, will do. (And yes, it's for receiving directly.)